azkaban-common/src/main/java/azkaban/imagemgmt/version/JdbcVersionSetLoader.java

```diff
@@ -0,0 +1,237 @@
+        throw new IOException(
+            "Expected only one VersionSet for versionSetMd5Hex " + versionSetMd5Hex);
+      }
+      if (versionSets.isEmpty()) {
+        logger.info("No VersionSet exists for versionSetMd5Hex " + versionSetMd5Hex);
+        return null;
+      }
+      VersionSet versionSet = versionSets.get(0);
+      this.idToVersionSet.put(versionSet.getVersionSetId(), versionSet);
+      this.md5ToVersionSet.put(versionSetMd5Hex, versionSet);
+      return versionSet;
+    } catch (SQLException e) {
+      logger.error("Failed to execute: " + SELECT_VSET_FROM_MD5, e);
+      throw new IOException("Failed to fetch VersionSet for versionSetMd5Hex " + versionSetMd5Hex,
+          e);
+    }
+  }
+
+  /**
+   * This method first check if the versionSet exists in the local copy of versionSets, otherwise,
+   * it gets the copy from the version_set table and updates the local copy.
+   */
+  @Override
+  public synchronized VersionSet getVersionSetById(int versionSetId) throws IOException {
+    if (this.idToVersionSet.containsKey(versionSetId)) {
+      return this.idToVersionSet.get(versionSetId);
+    }
+    final VersionSetHandler versionSetHandler = new VersionSetHandler();
+    try {
+      List<VersionSet> versionSets = this.dbOperator
+          .query(SELECT_VSET_FROM_ID, versionSetHandler, versionSetId);
+      if (versionSets.size() > 1) {
+        throw new IOException("Expected only one VersionSet for versionSetId " + versionSetId);
+      }
+      if (versionSets.isEmpty()) {
+        logger.info("No VersionSet exists for versionSetId " + versionSetId);
+        return null;
+      }
+      VersionSet versionSet = versionSets.get(0);
+      this.idToVersionSet.put(versionSetId, versionSet);
+      this.md5ToVersionSet.put(versionSet.getVersionSetMd5Hex(), versionSet);
+      return versionSet;
+    } catch (SQLException e) {
+      logger.error("Failed to execute: " + SELECT_VSET_FROM_ID, e);
+      throw new IOException("Failed to fetch VersionSet for versionSetId " + versionSetId, e);
+    }
+  }
+
+  /**
+   * Fetches all the rows from the version_set table and returns it as List of VersionSet.
+   */
+  @Override
+  public synchronized List<VersionSet> fetchAllVersionSets() throws IOException {
+    final VersionSetHandler versionSetHandler = new VersionSetHandler();
+    try {
+      return this.dbOperator.query(SELECT_ALL_VSET, versionSetHandler);
+    } catch (SQLException e) {
+      logger.error("Failed to execute: " + SELECT_ALL_VSET, e);
+      throw new IOException("Failed to fetch all VersionSets ", e);
+    }
+  }
+
+  /**
+   * A Handler class to modify the result set into List of VersionSet.
+   */
+  public static class VersionSetHandler implements ResultSetHandler<List<VersionSet>> {
+
+    @Override
+    public List<VersionSet> handle(ResultSet rs) throws SQLException {
+      List<VersionSet> versionSets = new ArrayList<>();
+      while (rs.next()) {
+        int versionSetId = rs.getInt(1);
```

Comment:
Would be nice to use mnemonics instead of hard-coded column numbers.

Reply:
done